### ISSUE TITLE
Number of Features to Integrate: Effect on UMAP Plot

### ISSUE THREAD

**Issue by MrModenait** [satijalab/seurat#3699]
Hello everybody, I am working on integrating four small datasets for clustering using the [standard workflow](https://satijalab.org/seurat/v3.2/integration.html). In order to perform certain analyses later, I have learnt how to manually select which features to integrate but also how to add genes for scaling. This has already been discussed in several other posts (https://github.com/satijalab/seurat/issues/1198, https://github.com/satijalab/seurat/issues/2270, https://github.com/satijalab/seurat/issues/2350, https://github.com/satijalab/seurat/issues/2590, amongst others) and is no new request.

However, although I am always using the same number of genes to build my anchors on (2000 genes in `FindIntegrationAnchors`) and the same variable features for ScaleData (`features = VariableFeatures(my_dataset)`), I seem to get different shapes on my UMAP plot depending on how many genes/features I integrated. Is this normal? Following [this link](https://imgur.com/a/2OztFH6) to Imgur you can see a plot created after running with 2000 integrated genes (`features.to.integrate = NULL`) and a plot created after running with all my genes (`features.to.integrate = background.genes_symbol` where `background.genes_symbol` is a list of every gene expressed at least once).

Thus the question is, since the only thing I change in my code is which **features I integrate** in the second block, why does calculating batch-corrected data for all my genes **alter the layout of my UMAP plot** even though **all other parameters** (anchors, genes being scaled and features being used for PCA) remain **unchanged**? 
For reference I will also add my code and a guide to interpret it. First I load my four objects which correspond to some cancer cells from bone marrow and primary tumour in two different cell lines:
```R
PALLAD_Bone <- readRDS("../Directory/Seurat Objects/PALLAD_Bone.rds")
PALLAD_Primary <- readRDS("../Directory/Seurat Objects/PALLAD_Primary.rds")
FORPA_Bone <- readRDS("../Directory/Seurat Objects/FORPA_Bone.rds")
FORPA_Primary <- readRDS("../Directory/Seurat Objects/FORPA_Primary.rds")
```
Afterwards I run the integration. In this case I have integrated all genes as you can see next to `features.to.integrate`, using `background.genes_symbol` which contains about 17000 genes.
```R
genes.in.use <- 2000
anchor.dims.in.use <- 30
PDX_Anchors <- FindIntegrationAnchors(object.list = list(PALLAD_Bone,
                                                         PALLAD_Primary,
                                                         FORPA_Bone,
                                                         FORPA_Primary), 
                                      dims = 1:anchor.dims.in.use,
                                      verbose = TRUE,
                                      anchor.features = genes.in.use)
background.genes_symbol <- readRDS(file = "../Directory/Gene Lists/Background genes (SYMBOL).rds")
PDX_Integrated <- IntegrateData(anchorset = PDX_Anchors,
                                features.to.integrate = background.genes_symbol,
                                dims = 1:30,
                                new.assay.name = "Integrated")
DefaultAssay(PDX_Integrated) <- "Integrated"
```
For brevity I rename the identities:
```R
Idents(PDX_Integrated) <- "orig.ident"
PDX_Integrated[["old.ident"]] <- Idents(object = PDX_Integrated)
PDX_Integrated <- RenameIdents(object = PDX_Integrated,
                               "FORPA (Bone Marrow)" = "FORPA (BM)",
                               "FORPA (Primary Tumour)" = "FORPA (PT)",
                               "PALLAD (Bone Marrow)" = "PALLAD (BM)",
                               "PALLAD (Primary Tumour)" = "PALLAD (PT)")
PDX_Integrated <- StashIdent(PDX_Integrated,save.name = "abbreviated.ident")
Idents(PDX_Integrated) <- "abbreviated.ident"
```
I then analyse cell cycle scores:
```R
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
PDX_Integrated <- CellCycleScoring(PDX_Integrated, 
                                   s.features = s.genes, 
                                   g2m.features = g2m.genes,
                                   set.ident = TRUE)
head(PDX_Integrated[[]])
```
Next, I find variable features manually in the *RNA assay* (I know IntegrateData does this for me but I want to add some cytokines of interest).
```R
#Find variable features
DefaultAssay(PDX_Integrated) <- "RNA"
genes.in.use <- 2000
PDX_Integrated <- FindVariableFeatures(PDX_Integrated, selection.method = "vst", nfeatures = genes.in.use)
DefaultAssay(PDX_Integrated) <- "Integrated"
```
I import my list of interesting cytokines, check that they are expressed in the dataset and then create a list of genes to scale
```R
total.genes <- list(rownames(PALLAD_Bone),
                    rownames(PALLAD_Primary),
                    rownames(FORPA_Bone),
                    rownames(FORPA_Primary))
common.genes <- Reduce(f = intersect, x = total.genes)
query.cytokines <- read.csv("../Directory/Query Cytokines.csv", encoding="UTF-8")
missing.cytokines <- setdiff(query.cytokines$X.U.FEFF.Gene.Name,common.genes)
write.csv(missing.cytokines,file = "Invalid Query Cytokines.csv")
query.genes <- read.csv("../Directory/Query genes.csv", encoding="UTF-8")
missing.genes <- setdiff(query.genes$X.U.FEFF.Gene.Name,common.genes)
write.csv(missing.genes,file = "Invalid Query genes.csv")
genes.to.scale <- list(VariableFeatures(PDX_Integrated),
                       query.cytokines$X.U.FEFF.Gene.Name,
                       query.genes$X.U.FEFF.Gene.Name)
genes.to.scale <- Reduce(f = union, x = genes.to.scale)
```
And then I scale.
```R
PDX_Integrated <- ScaleData(PDX_Integrated, features = genes.to.scale, vars.to.regress = c("nFeature_RNA","percent.mt","S.Score", "G2M.Score"), verbose = TRUE)
```
Finally, I run PCA analysis (only using variable features, not genes I have a personal interest in) and make myself a UMAP.
```R
#Run PCA using the variable features
PDX_Integrated <- RunPCA(PDX_Integrated,npcs = 30, verbose = TRUE, features = VariableFeatures(PDX_Integrated)) 
print(PDX_Integrated[["pca"]], dims = 1:20, nfeatures = 5)
VizDimLoadings(PDX_Integrated, dims = 1:2, reduction = "pca")
#Determine number of PCAs to use.
ElbowPlot(PDX_Integrated)
pca.chosen <- 16
DefaultAssay(PDX_Integrated) <- "Integrated"
PDX_Integrated <- RunUMAP(PDX_Integrated, reduction = "pca", dims = 1:pca.chosen)
```
So what could explain why the UMAP plot shifts? And which plot is a more accurate representation of my cell population?

**Comment by kleurless** [comment]
Hi,

Not an expert here, just sharing my views. I believe that it is normal that your UMAP plots shift when using a different set of features. UMAP calculations are determined with PCA dimensions. PCA is used to reduce dimensionality and sets your data variance "away" in dimensions. When you have more data (more features) to reduce dimensionality, your PCA dimensions should experience a little shift as well. So generally speaking, your input data for calculating the UMAP coordinates differs when using more features and so should your UMAP results.
Regarding your question of the most accurate representation kind of depends on your interest. I think that a smaller amount of features that you are interested in is preferable for analysis and reduces run time. Besides that, you keep all the features in the RNA assay so you can still plot them with FeaturePlot.

(BTW: make your issues shorter. This is waaaaay too long. You don't need to show so much code, just your question would be fine, you can always share your code if someone asks)

**Comment by MrModenait** [comment]
Thank you for your opinions @kleurless: I thought about this as well, but considering I have set `features = VariableFeatures(PDX_Integrated)` in `RunPCA()`, shouldn't the PCA results be independent of the other fifteen thousand features that I added in? I wondered whether these 2000 variable features would differ based on how many features were integrated, but the genes obtained from `FindVariableFeatures()` in the fifth block was identical regardless of whether all features or 2000 features were integrated. 
Regarding which approach is better, having integrated data for all genes would be useful as it would allow me to check e.g. expression of interesting cytokines mentioned in literature, or run a GSEA analysis for a cluster using [clusterProfiler](http://yulab-smu.top/clusterProfiler-book/chapter5.html#go-gene-set-enrichment-analysis). 
I take your point about shorter issues – I decided to share the code to enable the issue to be reproduced. 😀

**Comment by kleurless** [comment]
I must've missed that you set your features for PCA, that makes my point irrelevant. Then I'm unsure what is causing this behaviour, sure am curious tho!

**Comment by timoast** [comment]
UMAP is stochastic so will change slightly with different random seeds. Are the integration results identical in the two cases (for the genes that are common)? Are the PCA results identical?

I noticed you're defining cell cycle stage based on the integrated data, were all those cell cycle genes present in the features to be integrated in both cases? If not, I'd expect there to be differences just based on that, since you include cell cycle in `vars.to.regress` in `ScaleData`

**Comment by MrModenait** [comment]
> UMAP is stochastic so will change slightly with different random seeds. Are the integration results identical in the two cases (for the genes that are common)? Are the PCA results identical?
> 
> I noticed you're defining cell cycle stage based on the integrated data, were all those cell cycle genes present in the features to be integrated in both cases? If not, I'd expect there to be differences just based on that, since you include cell cycle in `vars.to.regress` in `ScaleData`

Dear Tim, thank you for your reply. I took a deeper look at the PCA results from four different possible scenarios (2000 genes versus all genes, cell cycle scoring on versus off) and it appears the answer to my question can be found in the cell cycle scoring. 
Indeed, the number of features I integrated affected how many cell cycle genes that `CellCycleScoring` could identify. With only two-thousand genes, nineteen cell cycle genes were not found in the integrated object, compared to three missing genes when all 22,788 genes were used. Consequently, the PCA results varied with the number of features when `CellCycleScoring` was on. When `CellCycleScoring` was disabled, however, the PCA results were independent of the number of features that was used. These findings were verified by running UMAP plots with a fixed seed.
In conclusion, I'll integrate all the features since I can identify more cell cycle genes that way. Thank you for your help!
